src/platform_impl/windows/event_loop.rs

```diff
@@ -1438,39 +1438,63 @@ unsafe fn public_window_callback_inner<T: 'static>(
+            let y = super::get_y_lparam(lparam as u32) as i32;
+            let position = PhysicalPosition::new(x as f64, y as f64);
 
-            if mouse_was_outside_window {
-                userdata.send_event(Event::WindowEvent {
-                    window_id: RootWindowId(WindowId(window)),
-                    event: CursorEntered {
-                        device_id: DEVICE_ID,
-                    },
-                });
+            'o: {
+                let mut w = userdata.window_state_lock();
+                let mouse_was_outside_window =
+                    !w.mouse.cursor_flags().contains(CursorFlags::IN_WINDOW);
 
-                // Calling TrackMouseEvent in order to receive mouse leave events.
-                unsafe {
-                    TrackMouseEvent(&mut TRACKMOUSEEVENT {
-                        cbSize: mem::size_of::<TRACKMOUSEEVENT>() as u32,
-                        dwFlags: TME_LEAVE,
-                        hwndTrack: window,
-                        dwHoverTime: HOVER_DEFAULT,
-                    })
+                let rect: RECT = unsafe {
+                    let mut rect: RECT = mem::zeroed();
+                    if GetClientRect(window, &mut rect) == false.into() {
+                        break 'o; // exit early if GetClientRect failed
+                    }
+                    rect
                 };
+
+                let x = (rect.left..rect.right).contains(&x);
+                let y = (rect.top..rect.bottom).contains(&y);
+
+                if mouse_was_outside_window && x && y {
+                    w.mouse
+                        .set_cursor_flags(window, |f| f.set(CursorFlags::IN_WINDOW, true))
+                        .ok();
+
+                    userdata.send_event(Event::WindowEvent {
+                        window_id: RootWindowId(WindowId(window)),
+                        event: CursorEntered {
+                            device_id: DEVICE_ID,
+                        },
+                    });
+
+                    // Calling TrackMouseEvent in order to receive mouse leave events.
+                    unsafe {
+                        TrackMouseEvent(&mut TRACKMOUSEEVENT {
+                            cbSize: mem::size_of::<TRACKMOUSEEVENT>() as u32,
+                            dwFlags: TME_LEAVE,
+                            hwndTrack: window,
+                            dwHoverTime: HOVER_DEFAULT,
+                        })
+                    };
+                } else if !(mouse_was_outside_window || x && y) {
+                    w.mouse
+                        .set_cursor_flags(window, |f| f.set(CursorFlags::IN_WINDOW, false))
+                        .ok();
+
+                    userdata.send_event(Event::WindowEvent {
+                        window_id: RootWindowId(WindowId(window)),
+                        event: CursorLeft {
+                            device_id: DEVICE_ID,
+                        },
+                    });
+                }
```

Comment:
Could you extract all of this mess into a function accepting: window, x, y, position and returning something like `PointerMoveKind { Move, Enter, Leave }` so you can then do something like that
```rust
match get_pointer_move_kind(window, position, x, y) {
    PointerMoveKind::Leave => {
        // Cursor left handling
    }
    PointerMoveKind::Enter => {
       // Cursor enter handling
    }
    PointerMoveKind::Move => (),
}